Move OMP/MKL thread initialization into ATen/Parallel

ilia-cher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#19011 Move OMP/MKL thread initialization into ATen/Parallel**

Summary:
This PR implements changes described in
https://github.com/pytorch/pytorch/issues/19001

Test:
BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake
./build/bin/thread_init_test

Differential Revision: [D14846034](https://our.internmc.facebook.com/intern/diff/D14846034)

Commit:
Move OMP/MKL thread initialization into ATen/Parallel

Summary:
This PR implements changes described in
https://github.com/pytorch/pytorch/issues/19001

Commit:
Update on "Move OMP/MKL thread initialization into ATen/Parallel"

Move OMP/MKL thread initialization into ATen/Parallel

Summary:
This PR implements changes described in
https://github.com/pytorch/pytorch/issues/19001

gh-metadata: pytorch pytorch 19011 gh/ilia-cher/1/head

ilia-cher commented:
will update to fix tbb_init_test

Commit:
Update on "Move OMP/MKL thread initialization into ATen/Parallel"

Move OMP/MKL thread initialization into ATen/Parallel

Summary:
This PR implements changes described in
https://github.com/pytorch/pytorch/issues/19001

gh-metadata: pytorch pytorch 19011 gh/ilia-cher/1/head

Commit:
Update on "Move OMP/MKL thread initialization into ATen/Parallel"

Move OMP/MKL thread initialization into ATen/Parallel

Summary:
This PR implements changes described in
https://github.com/pytorch/pytorch/issues/19001

gh-metadata: pytorch pytorch 19011 gh/ilia-cher/1/head

Commit:
Update on "Move OMP/MKL thread initialization into ATen/Parallel"

Move OMP/MKL thread initialization into ATen/Parallel

Summary:
This PR implements changes described in
https://github.com/pytorch/pytorch/issues/19001

gh-metadata: pytorch pytorch 19011 gh/ilia-cher/1/head

Commit:
Update on "Move OMP/MKL thread initialization into ATen/Parallel"

Move OMP/MKL thread initialization into ATen/Parallel

Summary:
This PR implements changes described in
https://github.com/pytorch/pytorch/issues/19001

gh-metadata: pytorch pytorch 19011 gh/ilia-cher/1/head

Commit:
Update on "Move OMP/MKL thread initialization into ATen/Parallel"

Move OMP/MKL thread initialization into ATen/Parallel

Summary:
This PR implements changes described in
https://github.com/pytorch/pytorch/issues/19001

gh-metadata: pytorch pytorch 19011 gh/ilia-cher/1/head

dzhulgakov commented on aten/src/ATen/Parallel.cpp:
```diff
@@ -0,0 +1,63 @@
+  }
+#ifdef _OPENMP
+  omp_set_num_threads(nthreads);
+  num_threads.store(nthreads);
```
shouldn't it be outside of #ifdef ?

ilia-cher replied:
we could put this outside, though it will never be used if there's no parallelization library available (get_num_threads will always return 1); it all becomes no-ops essentially

dzhulgakov commented:
should we rename it to get_intra_num_threads btw? there's already c10/ThreadPool so it's confusing a bit which one we're referring to where

ilia-cher commented:
@dzhulgakov I'm not against changing the name, though I thought of adding get_inter_op_num_threads later

That's related to ThreadPool - that's on the plan too, the plan is to basically move task launching into ATen/Parallel and use it instead of explicit thread pool usage.
We'll need to differentiate between unified thread pool and separate inter- and intra-op pools and add an API to control that (the idea is that, e.g. for TBB backend we don't have to use two separate pools and can just use TBB for everything, while for native backend we might have to use two pool at first).

We'll definitely have a chance to discuss this once again and change if needed once we get to inter-op parallelism PR.

soumith commented:
let do the name change later. We have to figure out what happens to the public interface `torch.get_num_threads / torch.set_num_threads` (and at:: equivalents) that is is userland right now.

Commit:
Update on "Move OMP/MKL thread initialization into ATen/Parallel"

Move OMP/MKL thread initialization into ATen/Parallel

Summary:
This PR implements changes described in
https://github.com/pytorch/pytorch/issues/19001

gh-metadata: pytorch pytorch 19011 gh/ilia-cher/1/head

Commit:
Update on "Move OMP/MKL thread initialization into ATen/Parallel"

Move OMP/MKL thread initialization into ATen/Parallel

Summary:
This PR implements changes described in
https://github.com/pytorch/pytorch/issues/19001

gh-metadata: pytorch pytorch 19011 gh/ilia-cher/1/head

Commit:
Update on "Move OMP/MKL thread initialization into ATen/Parallel"

Move OMP/MKL thread initialization into ATen/Parallel

Summary:
This PR implements changes described in
https://github.com/pytorch/pytorch/issues/19001

gh-metadata: pytorch pytorch 19011 gh/ilia-cher/1/head

Commit:
Update on "Move OMP/MKL thread initialization into ATen/Parallel"

Move OMP/MKL thread initialization into ATen/Parallel

Summary:
This PR implements changes described in
https://github.com/pytorch/pytorch/issues/19001

gh-metadata: pytorch pytorch 19011 gh/ilia-cher/1/head

ilia-cher commented:
(checked CI errors - infra errors, unrelated lint error in test_jit, and double checked that 
pytest -s -v test/test_namedtuple_return_api.py passes)

apaszke commented on aten/src/ATen/Parallel.cpp:
```diff
@@ -0,0 +1,63 @@
+// region might be different in the new thread;
+// Use init_num_threads() during thread initialization to ensure
+// consistent size of parallel region in different threads
+size_t get_num_threads() {
```
Are OpenMP settings thread-local? I'm pretty sure this PR doesn't guard all the places where we create threads in PyTorch (e.g. inside the `DataParallel` module).

ilia-cher replied:
The test explicitly demonstrates that setting number of threads in OMP does not affect number of threads in a newly created thread, so unfortunately we need to call at::init_num_threads when initializing new threads.

Regarding this PR - the purpose of this PR is to move existing thread setting code (in ATen/CPUGeneral and in THGeneral) in one place (Parallel.cpp) and not introduce new call sites. As you can see the two places where we explicitly initialized number of threads were torch module's init and autograd engine new threads.

Regarding `DataParallel` module, if we don't call `set_num_threads` (don't seem to do atm) in new worker threads then it can be an old bug that we should fix.
